### PR TITLE
MaxICACallsPerEpoch applied to processing DepositRecords of type STAKE

### PR DESCRIPTION
**Pull request by riley-stride**
## What is the purpose of the change

We limit the number of ICA calls that could process per epoch in order to avoid a potential scenario in which, after the channel having gone down for some time, the pending number of ICA transactions exceeds the number that can fit into a single block on the host zone and thus the batch ICA transaction fails. As the protocol is currently designed, without a threshold here, this error would require a software upgrade to recover from. 

## Brief Changelog

  - *Added MaxIcaCallsPerEpoch param in `x/stakeibc`*
  - *Added logic to cap number of ICA txs processed each epoch in `hooks.go:StakeExistingDepositsOnHostZones`*
  - *Added test `x/stakeibc/keeper/stake_existing_deposits_on_host_zones_test.go` (WIP)*
  - *Modified some scripts to manually test limiting the number of DRs processed. THESE NEED TO BE REVERTED BEFORE MERGING!*

## Testing and Verifying

This change added tests and can be verified as follows:
  - * (WIP) Adding unit test that adds more than `MaxIcaCallsPerEpoch` DepositRecords, then calls `StakeExistingDepositsOnHostZones` and verifies no more than `MaxIcaCallsPerEpoch` DRs are processed*
  - * (TODO) Add failure unit tests for `StakeExistingDepositsOnHostZones`

## Documentation and Release Note
- Changes made to scripts *MUST BE REVERTED BEFORE MERGING*. We can do this once the unit test works.

### CODE REVIEW

**Comment by riley-stride** [comment]
@sampocs requesting you here since the unit test in `stake_existing_deposits_on_host_zones_test.go` currently fails, as it invokes IBC functionality (`DelegateOnHost`), and I would like to ask your advice on how to either monkey patch that function or otherwise refactor in such a way that would allow us to properly test `StakeExistingDepositsOnHostZones`

**Review by sampocs — COMMENTED** [review]
Logic in StakeExistingDepositsOnHostZones looks good! Lemme think a little on how to handle the test and get back to you. 

Monkeypatching unfortunately didn't seem viable when I looked into it. The packages were very old, had no documentation and even the example in the README was calling functions that didn't exist. But I think we can restructure the code here!

File: x/stakeibc/keeper/stake_existing_deposits_on_host_zones_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		s.App.RecordsKeeper.AppendDepositRecord(s.Ctx, initialDepositRecord)
+	}
+	return
+}
+
+func (s *KeeperTestSuite) TestStakeExistingDepositsOnHostZonesSuccessful() {
+	s.SetupStakeExistingDepositsOnHostZones()
+
+	PrevDepositRecordCount := s.App.StakeibcKeeper.RecordsKeeper.GetDepositRecordCount(s.Ctx)
+	depositRecords := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx)
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/stake_existing_deposits_on_host_zones_test.go:54]: Agreed, good suggestion

File: x/stakeibc/keeper/stake_existing_deposits_on_host_zones_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	s.App.StakeibcKeeper.StakeExistingDepositsOnHostZones(s.Ctx, 100, depositRecords)
+	NewDepositRecordCount := s.App.StakeibcKeeper.RecordsKeeper.GetDepositRecordCount(s.Ctx)
+
+	expectedNumProcessedWithCap := s.App.StakeibcKeeper.GetParam(s.Ctx, stakeibc.KeyMaxICACallsPerEpoch)
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/stake_existing_deposits_on_host_zones_test.go:58]: Good idea, this allows us to iterate over several `MaxICACallsPerEpoch` param values and test them in sequence.

**Comment by asalzmann** [comment]
Sorry for the delay - will review this tomorrow!

**Review by asalzmann — APPROVED** [review]
Other ICA limits we should consider
- Unbonding sends [one ICA per Validator](https://github.com/Stride-Labs/stride/blob/21683499bdde58a53c1917e0e8d1ba65a4c88769/x/stakeibc/keeper/unbonding_records.go#L114) - if we have too many validators here, we might hit this issue (unlikely but possible?) 
- Rebalancing is limited by [maxNumRebalance](https://github.com/Stride-Labs/stride/blob/1496250eaaf2d7cd424f8f425d278ea7613a8de5/x/stakeibc/keeper/msg_server_rebalance_validators.go#L145)
- We send [one delegation / validator](https://github.com/Stride-Labs/stride/blob/1b56caafd8c7f662b5f49666113b6f36333203f6/x/stakeibc/keeper/msg_server_submit_tx.go#L104)

We should run some tests on what happens if we have ~100 validators we're delegating to on a host zone. [Ticket tracking this](https://stridelabs.atlassian.net/jira/software/projects/TEST/boards/2/backlog?selectedIssue=TEST-158)

**Comment by asalzmann** [comment]
What's the status of this PR? (not urgent, just want to make sure we don't lose the thread)

**Comment by asalzmann** [comment]
bumping (again, just so we don't forget this)

**Pull request closed without merging**